From 530d8882ca67caaaecbba72d26491dbf8f8f5e26 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Cl=C3=A9ment=20Le=20Goffic?= <clement.legoffic@foss.st.com>
Date: Tue, 21 May 2024 12:11:58 +0200
Subject: [PATCH 1152/1695] drivers: stm32_iwdg: add stop watchdog handler
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Implement .stop watchdog handler in order to allow the platform to
`halt` well.

The platform does not allow a way to stop an IWDG once started but a
solution exists.
We disable the platform reset ability of the IWDG we use and the IRQ
associated if the reset property is present in the DT.

Be careful, on STM32MP platforms, in A35TDCID, the system reset of the
IWDG1 cannot be disabled. Therefore, only the IWDG2 can be stopped.
The same behavior applies for IWDG3 in M33TDCID, you'll have to use the
IWDG4.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git d02655bfae87f2c7bb775b27610cc1df319a3754 drivers: stm32_iwdg: add stop watchdog handler]
Change-Id: I23138c723052824380d7354ed3b6043df23f9836
Signed-off-by: Clément Le Goffic <clement.legoffic@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/381299
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Reviewed-by: Antonio Maria BORNEO <antonio.borneo@foss.st.com>
Domain-Review: Antonio Maria BORNEO <antonio.borneo@foss.st.com>
---
 core/drivers/stm32_iwdg.c | 42 +++++++++++++++++++++++++++++++++++----
 1 file changed, 38 insertions(+), 4 deletions(-)

diff --git a/core/drivers/stm32_iwdg.c b/core/drivers/stm32_iwdg.c
index 8ccf1ef0c..599924246 100644
--- a/core/drivers/stm32_iwdg.c
+++ b/core/drivers/stm32_iwdg.c
@@ -6,6 +6,7 @@
 #include <assert.h>
 #include <drivers/clk.h>
 #include <drivers/clk_dt.h>
+#include <drivers/rstctrl.h>
 #include <drivers/wdt.h>
 #include <io.h>
 #include <keep.h>
@@ -87,6 +88,8 @@
  * @itr_chip - Interrupt chip device
  * @itr_num - Interrupt number for the IWDG instance
  * @itr_handler - Interrupt handler
+ * @reset - Reset controller device used to control the ability of the watchdog
+ *          to reset the system
  * @flags - Property flags for the IWDG instance
  * @timeout - Watchdog elaspure timeout
  * @hw_version - Watchdog HW version
@@ -101,6 +104,7 @@ struct stm32_iwdg_device {
 	struct itr_chip *itr_chip;
 	size_t itr_num;
 	struct itr_handler *itr_handler;
+	struct rstctrl *reset;
 	uint32_t flags;
 	unsigned long timeout;
 	unsigned int hw_version;
@@ -170,6 +174,16 @@ static TEE_Result iwdg_wait_sync(struct stm32_iwdg_device *iwdg)
 	return TEE_SUCCESS;
 }
 
+static void stm32_iwdg_it_ack(struct stm32_iwdg_device *iwdg)
+{
+	vaddr_t iwdg_base = get_base(iwdg);
+
+	if (iwdg->hw_version >= IWDG_ICR_MIN_VER)
+		io_setbits32(iwdg_base + IWDG_ICR_OFFSET, IWDG_ICR_EWIC);
+	else
+		io_setbits32(iwdg_base + IWDG_EWCR_OFFSET, IWDG_EWCR_EWIC);
+}
+
 static enum itr_return stm32_iwdg_it_handler(struct itr_handler *h)
 {
 	unsigned int __maybe_unused cpu = get_core_pos();
@@ -190,10 +204,7 @@ static enum itr_return stm32_iwdg_it_handler(struct itr_handler *h)
 	io_write32(iwdg_base + IWDG_KR_OFFSET, IWDG_KR_WPROT_KEY);
 
 	/* Disable early interrupt */
-	if (iwdg->hw_version >= IWDG_ICR_MIN_VER)
-		io_setbits32(iwdg_base + IWDG_ICR_OFFSET, IWDG_ICR_EWIC);
-	else
-		io_setbits32(iwdg_base + IWDG_EWCR_OFFSET, IWDG_EWCR_EWIC);
+	stm32_iwdg_it_ack(iwdg);
 
 	panic("Watchdog");
 
@@ -281,9 +292,27 @@ static void iwdg_wdt_start(struct wdt_chip *chip)
 	struct stm32_iwdg_device *iwdg = wdt_chip_to_iwdg(chip);
 
 	iwdg_start(iwdg);
+	if (iwdg->reset && iwdg->itr_handler)
+		stm32_iwdg_it_ack(iwdg);
 
 	if (configure_timeout(iwdg))
 		panic();
+
+	if (iwdg->reset)
+		if (rstctrl_assert(iwdg->reset))
+			panic();
+}
+
+static void iwdg_wdt_stop(struct wdt_chip *chip)
+{
+	struct stm32_iwdg_device *iwdg = wdt_chip_to_iwdg(chip);
+
+	if (iwdg->reset) {
+		if (rstctrl_deassert(iwdg->reset))
+			panic();
+		if (iwdg->itr_handler)
+			interrupt_disable(iwdg->itr_chip, iwdg->itr_num);
+	}
 }
 
 static void iwdg_wdt_refresh(struct wdt_chip *chip)
@@ -348,6 +377,7 @@ static TEE_Result iwdg_wdt_get_timeleft(struct wdt_chip *chip, bool *is_enabled,
 static const struct wdt_ops stm32_iwdg_ops = {
 	.init = iwdg_wdt_init,
 	.start = iwdg_wdt_start,
+	.stop = iwdg_wdt_stop,
 	.ping = iwdg_wdt_refresh,
 	.set_timeout = iwdg_wdt_set_timeout,
 	.get_timeleft = iwdg_wdt_get_timeleft,
@@ -387,6 +417,10 @@ static TEE_Result stm32_iwdg_parse_fdt(struct stm32_iwdg_device *iwdg,
 			return res;
 	}
 
+	res = rstctrl_dt_get_by_index(fdt, node, 0, &iwdg->reset);
+	if (res != TEE_SUCCESS && res != TEE_ERROR_ITEM_NOT_FOUND)
+		return res;
+
 	if (dt_info.status == DT_STATUS_OK_NSEC)
 		iwdg->flags |= IWDG_FLAGS_NON_SECURE;
 
-- 
2.47.2

